crossplane-capi-import/README.md

```diff
@@ -0,0 +1,257 @@
+use of composition functions) require the editing of specific blocks of the
+composition.
+
+The reason for this is that there is no looping inside the composition and
+details returned from the cloud do not directly correlate to information required
+by CAPI.
+
+The example of this inside EKS is the subnet groups required by the
+`AWSManagedControlPlane` resource which then require hard coded specification
+via patching.
+
+```yaml
+- fromFieldPath: status.subnetIds[0]
+  toFieldPath: spec.forProvider.manifest.spec.network.subnets[0].id
+- fromFieldPath: status.subnetIds[1]
+  toFieldPath: spec.forProvider.manifest.spec.network.subnets[1].id
+- fromFieldPath: status.subnetIds[2]
+  toFieldPath: spec.forProvider.manifest.spec.network.subnets[2].id
+- fromFieldPath: status.subnetIds[3]
+  toFieldPath: spec.forProvider.manifest.spec.network.subnets[3].id
+- fromFieldPath: status.subnetIds[4]
+  toFieldPath: spec.forProvider.manifest.spec.network.subnets[4].id
+- fromFieldPath: status.subnetIds[5]
+  toFieldPath: spec.forProvider.manifest.spec.network.subnets[5].id
+```
+
+It may be feasible to correct this through the use of composition functions but
+this has not yet been explored as an option.
+
+### CAPI resources
+
+Compositions implementing this RFC will create the following CAPI resources:
+
+- Cluster
+- MachinePool
+
+Additionally cloud specific providers should be created
+
+| AWS                    | Azure                    | GCP                    |
+|------------------------|--------------------------|------------------------|
+| AWSManagedCluster      | AzureManagedCluster      | GCPManagedCluster      |
+| AWSManagedControlPlane | AzureManagedControlPlane | GCPManagedControlPlane |
+| AWSManagedMachinePool  | AzureManagedMachinePool  | GCPManagedMachinePool  |
+
+### `kubeconfig`
+
+The `kubeconfig` secret is normally generated by CAPI controllers and stored
+in the cluster namespace as `<cluster_name>-kubeconfig`
+
+As CAPI should be running in a non-reconcilling mode for imported clusters,
+it is not known if this kubeconfig will be generated automatically by CAPI or
+whether it needs to be imported separately.
+
+To work around potential limitations here, crossplane leverages the
+`cluster-auth` capabilities and stores this secret as `<cluster_name>-kubeconfig-cluster-auth`
+
+This secret contains a slightly different structure from CAPI and places the
+kubeconfig at the secret data key `data.kubeconfig`.
+
+This cannot be controlled via crossplane and in order to make this secret
+compatible with our own deployments via `App-Platform` we may need to enhance
+our `App` CR to accept a secret data key, or include this location in the list
+of key locations checked for a `kubeconfig` value.
+
+### Discovery
+
+With the introduction of EKS/AKS/GKE clusters to CAPI management clusters, we
+need to be able to differentiate between the different types and usage.
+
+This can be achieved through the use of the type of cluster defined as
```

Comment:
why do we need the cluster type to determine whether a cluster was imported? we only need `k8s.io/managed-by: crossplane`, right?

Reply:
Using AWS/EKS clusters as the example, the default check to determine if a cluster is standard CAPA or CAPA EKS would be to use the `infrastructureRef` as the type is different between these, however the `infrastrucureRef` kind would be the same between a fully managed CAPA EKS cluster and an imported EKS cluster.

Simply checking for the presence or absence of the annotation only tells you if the cluster is adopted or not but doesn't allow for separation between fully managed EKS and a standard CAPI cluster.

The inclusion of the table below was to specifically highlight how a CAPI cluster would appear in each of the given scenarios and how you can differentiate between what's fully managed and what's imported.

Reply:
gotcha, nice